test/opt/test_set_spec_const_default_value.cpp

```diff
@@ -0,0 +1,165 @@
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "pass_fixture.h"
+
+namespace {
+using namespace spvtools;
+
+using SpecIdToValueMap = opt::SetSpecConstantDefaultValuePass::SpecIdToValueMap;
+
+struct SetSpecConstantDefaultValueTestCase {
+  const char* code;
+  SpecIdToValueMap default_values;
+  const char* expected;
+};
+
+using SetSpecConstantDefaultValueParamTest =
+    PassTest<::testing::TestWithParam<SetSpecConstantDefaultValueTestCase>>;
+
+TEST_P(SetSpecConstantDefaultValueParamTest, TestCase) {
+  const auto& tc = GetParam();
+  SinglePassRunAndCheck<opt::SetSpecConstantDefaultValuePass>(
+      tc.code, tc.expected, /* skip_nop = */ false, tc.default_values);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    Types, SetSpecConstantDefaultValueParamTest,
+    ::testing::ValuesIn(std::vector<SetSpecConstantDefaultValueTestCase>{
+        // 0. empty
+        {"", SpecIdToValueMap{}, ""},
+        // 1. empty with non-empty values to set
+        {"", SpecIdToValueMap{{1, {100}}, {2, {100, 200}}}, ""},
+        // 2. bool type
+        {
+          // code
+          "OpDecorate %1 SpecId 100\n"
+          "%bool = OpTypeBool\n"
+          "%1 = OpSpecConstantTrue %bool\n",
+          // default values
+          SpecIdToValueMap{{100, {0}}},
+          // expected
+          "OpDecorate %1 SpecId 100\n"
+          "%bool = OpTypeBool\n"
+          "%1 = OpSpecConstantFalse %bool\n",
+        },
+        // 3. 32-bit int type
+        {
+          // code
+          "OpDecorate %1 SpecId 100\n"
+          "%int = OpTypeInt 32 1\n"
+          "%1 = OpSpecConstant %int 100\n",
+          // default values
+          SpecIdToValueMap{{100, {0x7fffffff}}},
+          // expected
+          "OpDecorate %1 SpecId 100\n"
+          "%int = OpTypeInt 32 1\n"
+          "%1 = OpSpecConstant %int 2147483647\n",
+        },
+        // 4. 64-bit uint type
+        {
+          // code
+          "OpDecorate %1 SpecId 100\n"
+          "%ulong = OpTypeInt 64 0\n"
+          "%1 = OpSpecConstant %ulong 100\n",
+          // default values
+          SpecIdToValueMap{{100, {0xffffffff, 0xffffffff}}},
```

Comment:
Make the two words different so you know you've put them in the right order.  Suggest subtracing 1 from the first word.
That should subtract one from the decimal result.